File: src/ca/mindmagic/game/map/grid/pattern/Hexagon.java
package ca.mindmagic.game.map.grid.pattern;

import java.awt.Point;
import java.util.Arrays;

public class Hexagon extends Shape{

  // HEIGHT = shortest distance between the center and any of the hex sides
  // where sideLength is equal to one.
  public static final float HEIGHT = (float)Math.sqrt(0.75);

  public Hexagon(){
    super();
  }

  public Hexagon(int sideLength){
    super(sideLength);
  }

  @Override public int numberOfSides() {
    return 6;
  }

  @Override public int lengthOfSides() {
    return sideLength;
  }

  @Override public Point[] getVertices() {
    return getVertices(sideLength);
  }

  public static Point[] getVertices(int sideLength) {
    Point[] vertices = new Point[6];
    vertices[0] = topPoint(sideLength);
    vertices[1] = upperRightPoint(sideLength);
    vertices[2] = lowerRightPoint(sideLength);
    vertices[3] = bottomPoint(sideLength);
    vertices[4] = lowerLeftPoint(sideLength);
    vertices[5] = upperLeftPoint(sideLength);
    return vertices;
  }

  public Double[] getVerticesAsDoubleArray(double scalar){
    Double[] vertices = new Double[12];
    return vertices;
  }

  private static Double[] scale(Double[] values, double scalar){
    return Arrays.stream(values).map(x -> x*scalar)
        .toArray(Double[]::new);
  }

  private static Point upperRightPoint(int size) {
    int x, y;
    x = (int)(HEIGHT * size);
    y = -size/2;
    return new Point(x,y);
  }

  private static Point lowerRightPoint(int size) {
    int x, y;
    x = (int)(HEIGHT * size);
    y = size/2;
    return new Point(x,y);
  }

  private static Point bottomPoint(int size) { // done
    int x, y;
    x = 0;
    y = size;
    return new Point(x,y);
  }

  private static  Point lowerLeftPoint(int size) {
    int x, y;
    x = -(int)(HEIGHT * size);
    y = size/2;
    return new Point(x,y);
  }

  private static Point upperLeftPoint(int size) {
    int x, y;
    x = -(int)(HEIGHT * size);
    y = -size/2;
    return new Point(x,y);
  }

  private static Point topPoint(int size) { // done
    int x, y;
    x = 0;
    y = -size;
    return new Point(x,y);
  }

  public String toString(){
    return getVertices().toString();
  }


}

File: src/ca/mindmagic/game/map/grid/pattern/Hexagon.kt
package ca.mindmagic.game.map.grid.pattern

import javafx.geometry.Point2D
import java.util.Arrays

class Hexagon(sideLength: Double) : Shape(sideLength) {

    override fun numberOfSides(): Int {
        return 6
    }

    override fun getVertices() = getVertices(sideLength)

    companion object {

        // HEIGHT = shortest distance between the center and any of the hex sides
        // where sideLength is equal to one.
        val HEIGHT = Math.sqrt(0.75)

        fun getVertices(sideLength: Double): Array<Point2D> {
            return arrayOf(
                    topPoint(sideLength),
                    upperRightPoint(sideLength),
                    lowerRightPoint(sideLength),
                    bottomPoint(sideLength),
                    lowerLeftPoint(sideLength),
                    upperLeftPoint(sideLength))
        }

        private fun upperRightPoint(size: Double): Point2D {
            val x = (HEIGHT * size)
            val y = -size / 2.0
            return Point2D(x, y)
        }

        private fun lowerRightPoint(size: Double): Point2D {
            val x = (HEIGHT * size)
            val y = size / 2.0
            return Point2D(x, y)
        }

        private fun bottomPoint(size: Double): Point2D { // done
            val x = 0.0
            val y = size
            return Point2D(x, y)
        }

        private fun lowerLeftPoint(size: Double): Point2D {
            val x = -(HEIGHT * size)
            val y = size / 2.0
            return Point2D(x, y)
        }

        private fun upperLeftPoint(size: Double): Point2D {
            val x = -(HEIGHT * size)
            val y = -size / 2.0
            return Point2D(x, y)
        }

        private fun topPoint(size: Double): Point2D { // done
            val x = 0.0
            val y = -size
            return Point2D(x, y)
        }
    }


}
